### PR TITLE
Added support for Jupyter Notebooks.

### PR DESCRIPTION
**Pull request by jrobsondev**
I have added support for Jupyter Notebooks natively rather than having to download an additional tool. I figured this could make the tool easier to use for students. I have included a sample file that uses the same python that is in original sample file.

Couple of things:
- I have included an env folder and launch.json exclusion in the .gitignore. This can be removed, however, thought it may be useful for anybody using VSCode that may contribute.
- I have not updated the README. Let me know if you would like this to be done as well.

### CODE REVIEW

File: allowed.py [hunk 1]
```diff
@@ -558,6 +559,10 @@ def check_tree(tree: ast.AST, constructs: tuple, source: list) -> list:
             del errors[index]
     return errors
 
+def is_filetype_supported(filename: str) -> bool:
```
Comment (mwermelinger) [inline allowed.py:562]: The code can be simplified by omitting this function. See my later comment.

File: allowed.py [hunk 2]
```diff
@@ -581,7 +586,10 @@ def check_file(filename: str, constructs: tuple) -> None:
     """Check that the file only uses the allowed constructs."""
     try:
         with open(filename) as file:
-            source = file.read()
+            if file.name.endswith(".ipynb"):
```
Comment (mwermelinger) [inline allowed.py:589]: Why file.name instead of filename?

File: allowed.py [hunk 3]
```diff
@@ -610,6 +618,17 @@ def check_file(filename: str, constructs: tuple) -> None:
         else:
             print(f"{filename}: can't parse: {message}")
 
+def read_jupyter_notebook(file_contents: str) -> str:
+    """Returns a string representation of all code cells in a Jupyter Notebook"""
+    jobject = json.loads(file_contents)
+    code_blocks = [x for x in jobject['cells'] if x['cell_type'] == 'code']
```
Comment (mwermelinger) [inline allowed.py:624]: Creating a list of all code cells uses extra memory. I'd suggest to add an if to the for-loop below:
```py
for cell in jobject...:
    if cell['cell_type'] == ...:
        for line in cell['source']:
            ...
```

File: allowed.py [hunk 4]
```diff
@@ -640,7 +659,7 @@ def check_file(filename: str, constructs: tuple) -> None:
     for name in args.file_or_folder:
         if os.path.isdir(name):
             check_folder(name, args.unit)
-        elif name.endswith(".py"):
+        elif is_filetype_supported(name):
```
Comment (mwermelinger) [inline allowed.py:662]: It's simpler and faster to check if name ends with py or ends with ipynb than calling a function that splits the pathname.

**Comment by mwermelinger** [comment]
Many thanks for this contribution. Good idea to not require `nbqa`! It's fine to add things to .gitignore. Please do edit the README. The option of using with nbqa should remain, as it gives error messages relative to the cells. However, the first option (of saving first to a .py file), can be replaced by a mention that it now works directly on notebooks.

BTW, are you an M269 student or did you come across `allowed` in some other way?

File: allowed.py (hunk 1 above)
Comment (jrobsondev) [reply allowed.py:562]: Removed.

File: allowed.py (hunk 2 above)
Comment (jrobsondev) [reply allowed.py:589]: No reason, good spot!

File: allowed.py (hunk 4 above)
Comment (jrobsondev) [reply allowed.py:662]: That's fair enough. I switched for the function as I thought it read better than `elif name.endswith("py") or name.endswith("ipynb")` but didn't really consider the performance implicateions. Have switched back.

**Comment by jrobsondev** [comment]
Thanks for taking the time to look over my request. I have made the changes as per your comments above. I have also updated the README as discussed.

Let me know if there are any further changes.

> BTW, are you an M269 student or did you come across allowed in some other way?

I'm an M269 student and `allowed` saved me from dropping marks on a TMA I just submitted. I thought it was a useful tool and figured it could be made a bit more user-friendly by dropping the need to exportto `.py` or download another piece of software to work with the default TMA format.

**Comment by mwermelinger** [comment]
This has now been merged. Thanks again.

**Comment by mwermelinger** [comment]
I'm going to try submit a short description of `allowed` to a conference and I'd like to acknowledge your contribution. Can you let me know your name? You can do it on the M269 technical forum instead of publicly here. Thanks.

**Comment by jrobsondev** [comment]
Hi Michel,

I have sent you an email privately.

**Pull request closed without merging**